include/mpm_base.tcc

```diff
@@ -0,0 +1,628 @@
+
+      // Add material to list
+      auto result = materials_.insert(std::make_pair(mat->id(), mat));
+
+      // If insert material failed
+      if (!result.second) {
+        status = false;
+        throw std::runtime_error(
+            "New material cannot be added, insertion failed");
+      }
+    }
+  } catch (std::exception& exception) {
+    console_->error("#{}: Reading materials: {}", __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Apply nodal tractions
+template <unsigned Tdim>
+bool mpm::MPMBase<Tdim>::apply_nodal_tractions() {
+  bool status = true;
+  try {
+    // Read and assign nodes tractions
+    if (!io_->file_name("nodal_tractions").empty()) {
+      // Get mesh properties
+      auto mesh_props = io_->json_object("mesh");
+      // Get Mesh reader from JSON object
+      const std::string reader =
+          mesh_props["mesh_reader"].template get<std::string>();
+      // Create a mesh reader
+      auto node_reader =
+          Factory<mpm::ReadMesh<Tdim>>::instance()->create(reader);
+
+      bool nodal_tractions =
+          mesh_->assign_nodal_tractions(node_reader->read_particles_tractions(
+              io_->file_name("nodal_tractions")));
+      if (!nodal_tractions)
+        throw std::runtime_error("Nodal tractions are not properly assigned");
+    } else
+      nodal_tractions_ = false;
+  } catch (std::exception& exception) {
+    console_->error("#{}: Nodal traction: {}", __LINE__, exception.what());
+    status = false;
+    nodal_tractions_ = false;
+  }
+  return status;
+}
+
+//! Apply properties to particles sets (e.g: material, remove step)
+template <unsigned Tdim>
+bool mpm::MPMBase<Tdim>::apply_properties_to_particles_sets() {
+  bool status = false;
+  // Assign material to particle sets
+  try {
+    // Get particle properties
+    auto particle_props = io_->json_object("particle");
+    // Get particle sets properties
+    auto particle_sets = particle_props["particle_sets"];
+    // Iterate over each particle sets
+    for (const auto& psets : particle_sets) {
+      // Assign material to particle sets
+      // Get set material from list of materials
+      auto set_material = materials_.at(psets["material_id"]);
+      // Get sets ids
+      std::vector<unsigned> sids = psets["set_id"];
+      // Assign material to particles in the specific sets
+      for (const auto& sitr : sids) {
+        mesh_->iterate_over_particle_set(
+            sitr, std::bind(&mpm::ParticleBase<Tdim>::assign_material,
+                            std::placeholders::_1, set_material));
+      }
+
+      // Assign remove steps to particle sets
+      // Check if remove step is needed
+      if (psets["remove"] == true) {
+        // Get the number of remove step
+        mpm::Index rstep = psets["rstep"];
+        // Get ids of sets need to be removed
+        std::vector<unsigned> sids = psets["set_id"];
```

Comment:
Remove get ids from here as it is already implemented above. For each `particle_set`, read properties. If `remove` is set to true, call `mesh_->create_remove_step(rstep, sid);`